larpandoracontent/LArTwoDReco/TrackInEMShowerAlgorithm.cc

```diff
@@ -0,0 +1,892 @@
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+void TrackInEMShowerAlgorithm::GetExtrapolatedCaloHits(const ClusterAssociation &clusterAssociation, const ClusterList *const pClusterList,
+    CaloHitVector &extrapolatedCaloHitVector, ClusterToCaloHitListMap &clusterToCaloHitListMap) const
+{
+    const CartesianVector &upstreamPoint(clusterAssociation.GetUpstreamMergePoint()), &downstreamPoint(clusterAssociation.GetDownstreamMergePoint());
+    const float minX(std::min(upstreamPoint.GetX(), downstreamPoint.GetX())), maxX(std::max(upstreamPoint.GetX(), downstreamPoint.GetX()));
+
+    // ATTN: Consider hits from merging clusters as merging point may not be cluster end but do not put them in clusterToCaloHitListMap
+    for (const Cluster *const pCluster : *pClusterList) 
+    {
+        const OrderedCaloHitList &orderedCaloHitList(pCluster->GetOrderedCaloHitList());
+        for (const OrderedCaloHitList::value_type &mapEntry : orderedCaloHitList)
+        {
+            for (const CaloHit *const pCaloHit : *mapEntry.second) 
+            {
+                const CartesianVector &hitPosition(pCaloHit->GetPositionVector());
+                if ((hitPosition.GetX() < minX) || (hitPosition.GetX() > maxX) || (hitPosition.GetZ() < upstreamPoint.GetZ()) || (hitPosition.GetZ() > downstreamPoint.GetZ()))
+                    continue;
+                
+                const float distanceFromLine(clusterAssociation.GetConnectingLineDirection().GetCrossProduct(hitPosition - upstreamPoint).GetMagnitude());
+                if (distanceFromLine > m_distanceFromLine)
+                    continue;
+                
+                extrapolatedCaloHitVector.push_back(pCaloHit);
+
+                if ((pCluster != clusterAssociation.GetUpstreamCluster()) && (pCluster != clusterAssociation.GetDownstreamCluster()))
+                    clusterToCaloHitListMap[pCluster].push_back(pCaloHit);
+            }
+        }
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TrackInEMShowerAlgorithm::IsTrackContinuous(const ClusterAssociation &clusterAssociation, CaloHitVector &extrapolatedCaloHitVector) const
+{
+    const CartesianVector &upstreamMergePoint(clusterAssociation.GetUpstreamMergePoint()), &downstreamMergePoint(clusterAssociation.GetDownstreamMergePoint());
+    const CartesianVector &trackDirection(clusterAssociation.GetConnectingLineDirection());
+    const CartesianVector trackStep(trackDirection * m_lineSegmentLength);
+
+    // To handle final segment - merge track remainder with preceding segment and if track remainder was more than half trackstep split into two
+    const float trackLength((downstreamMergePoint - upstreamMergePoint).GetMagnitude());
+    const unsigned int fullSegments(floor(trackLength / m_lineSegmentLength));
```

Comment:
It is not, but i'll add in this check. m_lineSegmentLength is set by the user, and so if they set it to be 0 then this will always fail. If we raised an exception for an invalid parameter then this would be raised for all events, until the user changed the parameter. So is it better to return false here (in which case the alg will do nothing) with a helpful print statement rather than throwing an exception?

Reply:
If you add this check in ReadSettings, and print a useful message before returning e.g. STATUS_CODE_INVALID_PARAMETER, then the whole job will actually refuse to start until the parameter is changed. As things stand, there's no other code that uses this member variable, so it would then be guaranteed not to be zero before use in division. But, future developments may change this, so probably good to protect in this function too, just before the division, e.g. raising an exception if negligible difference to zero `(std::fabs(m_lineSegmentLength) < std::numeric_limits<float>::epsilon())`